### ISSUE TITLE
sparse tensor eliminate_zeros

### ISSUE THREAD

**Issue by monchin** [pytorch/pytorch#31742]
## 🚀 Feature
In `scipy.sparse.coo_matrix`, there is a method [`eliminate_zeros`](https://github.com/scipy/scipy/blob/v1.4.1/scipy/sparse/coo.py#L556-L564). It eliminates all the zero values in `values` and the corresponding indices in `indices`, which I think simple but sometimes  useful.

I tried to  implement it myself for pytorch, but I found that `indices` and `values` are both private attributes for sparse tensor, and I didn't find any `set` functions for them. Meanwhile, if we get the `indices` and `values`, and eliminate zeros, and then create a new sparse tensor with eliminated ones, this way works but in my opinion inefficient. So I think maybe we need to make method `eliminate_zeros` in the sparseTensor class.

**Comment by mrshenli** [comment]
@monchin Does the following work for you?

```python
i = torch.LongTensor([[0, 2], [1, 0], [1, 2]])
v = torch.FloatTensor([3,      4,      0    ])
x = torch.sparse.FloatTensor(i.t(), v, torch.Size([2,3]))

def eliminate_zeros(x):
    mask = x._values().nonzero()
    nv = x._values().index_select(0, mask.view(-1))
    ni = x._indices().index_select(1, mask.view(-1))
    return torch.sparse.FloatTensor(ni, nv, torch.Size([2,3]))

eliminate_zeros(x)
```

**Comment by monchin** [comment]
> @monchin Does the following work for you?
> 
> ```python
> i = torch.LongTensor([[0, 2], [1, 0], [1, 2]])
> v = torch.FloatTensor([3,      4,      0    ])
> x = torch.sparse.FloatTensor(i.t(), v, torch.Size([2,3]))
> 
> def eliminate_zeros(x):
>     mask = x._values().nonzero()
>     nv = x._values().index_select(0, mask.view(-1))
>     ni = x._indices().index_select(1, mask.view(-1))
>     return torch.sparse.FloatTensor(ni, nv, torch.Size([2,3]))
> 
> eliminate_zeros(x)
> ```

Thank you for your reply, I'm now using this method by createing a new sparse tensor, but I think it is inefficient. Maybe operating private members `indices` and `value` would be a more efficient and elegant way.

**Comment by cpuhrsch** [comment]
@monchin 

What is your use case for this? What do you intend to do with this at a higher level?

**Comment by jdavidls** [comment]
I miss `sparse_tensor.coalesce(eliminate_zeros=True)`.

**Comment by cpuhrsch** [comment]
@jdavidls - what do you need it for?

**Comment by pearu** [comment]
Btw, here is a method to eliminate existing zeros (requires the current main branch):
```python
torch._masked._combine_input_and_mask(sum, sparse_tensor, sparse_tensor)
```

Example:
```python
>>> t = torch.tensor([[0, 1], [2, 0]]).to_sparse()
>>> t
tensor(indices=tensor([[0, 1],
                       [1, 0]]),
       values=tensor([1, 2]),
       size=(2, 2), nnz=2, layout=torch.sparse_coo)
>>> t._values()[0] = 0
>>> torch._masked._combine_input_and_mask(sum, t, t)
tensor(indices=tensor([[1],
                       [0]]),
       values=tensor([2]),
       size=(2, 2), nnz=1, layout=torch.sparse_coo)

```

**Comment by cpuhrsch** [comment]
@pearu - but that's all private functionality which we don't expose to the public

**Comment by pearu** [comment]
> but that's all private functionality which we don't expose to the public

Yes, indeed. My example may be useful only for defining/testing the semantics of `eliminate_zeros`. For instance, the implementation in https://github.com/pytorch/pytorch/issues/31742#issuecomment-570046628 is broken because it determines the mask using uncoaleased data. For example, if the uncoalesced data contains values 1 and -1 for the same index, the corresponding element is not eliminated and it will show up as 0 when coalescing the tensor.

**Comment by sambaPython24** [comment]
I would like to have that function as well. It would be particularly interesting combined with a ```torch.where``` function to eliminate entries in a certain range.

**Comment by Filco306** [comment]
Am very interested in this functionality as well, preferably a memory-efficient one.

**Comment by ClaudMor** [comment]
I'd be interested in such a method too.

A possible application would be to replicate the removal of reciprocal edges from graphs before GNN training as shown [here](https://github.com/deezer/gravity_graph_autoencoders/blob/fa098943ad8b25de14882d76f3ad08253b30ab35/gravity_gae/preprocessing.py#L170). Some graphs are too big to be represented via a dense adjacency matrix, hence the sparsity requirement.
